### DISCUSSION TITLE
Using sixels in prompts (CMD, PowerShell, and Bash / WSL)

### DISCUSSION THREAD

**Discussion by PhMajerus** [microsoft/terminal#17876] [Show and tell]
# Using sixels in prompts

This article shows how to embed sixels in string literals to include an image in the prompt in the Command Prompt (cmd.exe), Windows PowerShell (5.x), and PowerShell (6.x and later, pka PS Core).

## Command Prompt (cmd.exe)

Here is a `PROMPT` command for cmd.exe to include the MS-DOS logo ![image](https://github.com/user-attachments/assets/b0cb8a41-1b48-48fb-a77d-7a688080aaf6) as a prefix.

```batch
prompt $EP0;1q^"1;1;30;20#0;2;0;0;0#1;2;97;0;0#2;2;50;0;0#3;2;75;0;97#4;2;37;0;63#5;2;97;97;0#6;2;50;50;0#0~^!6@`XDA`PPp@xDA`PPPpP`AACw$$#1?}}}^!4MA???__#5???w{]^!6M[[G$$#2^!4?oooOCA$$#3^!8?_w[^!5MA^!4?__#6??O__o$$#4^!10?_O???oCA??_-#0~???~@@~??_~??~@Bcg~???~aADDHo$$#1?~~~^!8?~~#5???@B^!5?K[wwo$$#3^!8?~~#4^^^!9?~#3~~$$#6^!17?AC^!5?O_-#0~???BAAABA@wGEBANO_?@ABAA@??_^^$$#1?~^^^^^!6[MF@#5^!4?@FNM^!4[]NFB$$#2??^!8_o?E@#6???MWoo^!5_ow[$$#3^!9?@??OW[[O^!4?@$$#4^!12?^!6_-#0@^!9A@@^!7A@^!6A@@$$#2?^!9@#4??^!7@#6?^!6@$E\$E[3C$S$P$G$S
```

![image](https://github.com/user-attachments/assets/437f97f7-e9bc-4b30-9937-c040c6454292)

You can create a batch file containing `@echo off` on the first line and that command on the second line, and provide its path to the Command Processor in the registry at `HKCU\Software\Microsoft\Command Processor\AutoRun` (`REG_SZ` or `REG_EXPAND_SZ`). This will make that command run automatically when launching cmd.exe.

## PowerShell

Here is a replacement `prompt` function for PowerShell that replaces the text `PS` prefix with the modern (black) PowerShell logo ![image](https://github.com/user-attachments/assets/6e6b003e-f958-466b-87c0-d62bc2625afd) using Sixels.

```powershell
function prompt { return "`eP0;1q`"1;1;26;20#1;2;89;91;91#2;2;71;77;83#3;2;60;64;77#4;2;50;61;75#5;2;33;42;56#6;2;16;24;36#7;2;13;19;30#8;2;8;14;20#9;2;5;8;11#10;2;2;2;3#1???oE???Ooo_#7?ow[[CC#4!5AC`$#3!4?O?!13A!5?G`$#2!4?GA???G#8!5?__ww{{[CS#2q`$#5!4?_C#4??g#9!12?_wG`$#6!5?w{{CCK[{KC!8?_-#5???O@!4?_!4?_W!6?O@`$#1??wB!6?``b}{W#9?ow{}NFF`$#6???_}~^NDAO#8???@fNFB@#10ow#4_A`$???G!5?@#2AS@A!9?[@`$#3???C#7??_oy[KG?@A!7?G`$#3!14?C-#5??C??C?O??CAK!5?C??G`$#3?_@???O?@???@!5?G??O`$#1?E!4?KMEBB@?!5K#2???_N`$?W!7?C#9_oo_aBBB@?_`$#6??GJ@OA??G???POOO!4?C`$#7??os}B@@O???A!4?O`$#4??A??G??G#10!6?___a~^@`$#8!5?!4_oWK?A@???O??A-#2?@#3@@#4!16@#3@#2@`e\`e[3C$($executionContext.SessionState.Path.CurrentLocation)$('>' * ($nestedPromptLevel + 1)) " }
```

![image](https://github.com/user-attachments/assets/1b8c0ed4-cf54-43cb-b4a9-36982660d245)

Of course, you can include that line in your `$profile` file to do it automatically when launching PowerShell.

## Windows PowerShell

Here is the equivalent for the legacy (blue) Windows PowerShell logo ![image](https://github.com/user-attachments/assets/f4b2107e-be8e-48f2-9473-17e63fef10b4).

```powershell
function prompt { return "$([char]0x1B)P0;1q`"1;1;26;20#1;2;25;44;78#2;2;25;45;83#3;2;28;50;88#4;2;28;52;91#5;2;30;55;96#6;2;64;75;93#7;2;89;91;94#8;2;27;45;80#9;2;42;56;85#10;2;20;39;67#1???oE#5{{{CCK[C#2!6?_wWKC#5_`$#4!4?o!7?w{{{[CC!5?Y`$#2!4?GA#3!10Aayy]EEAAC`$#7!8?Owo_#8!9?_ow`$#6!8?g-#4???G?ow{{[WG?@@!8?E`$#1??G@#9!5?_!5?G#2?MF@??O`$#2??_A#6!5?@A?@?C#1_!4?owC#5W@`$#8??O#5o~NFBBAC#3???AFF@!4?_@`$???C#7!6?``v}}W#8Owow}NFJ`$#10!14?_-#4??}}~R@@#1?G??A???__a~vB`$#8?E!5?_oow}oabbBB@??C`$#3?o@??__#10O??C??P!5O?G#4O@`$#2?G#5?@#6?KO?@C??L!5?G#5??_M`$#9!6?A?G!9?C#3??G`$#7!6?KMEBB@?!5K-#3?!20@#5@$([char]0x1B)\$([char]0x1B)[3C$($executionContext.SessionState.Path.CurrentLocation)$('>' * ($nestedPromptLevel + 1)) " }
```

![image](https://github.com/user-attachments/assets/975a3ecb-d18e-4baa-b715-203b1f4f2165)

## Bash (WSL)

Finally, here is how to embed a sixel logo in a Bash prompt, using the Ubuntu logo ![image](https://github.com/user-attachments/assets/cee15d77-7e54-4275-bbf8-c445b6da92da).

```bash
export PS1='\eP0;1q"1;1;20;20#1;2;91;25;0#2;2;91;30;3#3;2;91;31;9#4;2;91;22;0#5;2;94;61;55#6;2;91;33;17#7;2;97;93;91#1??__?KCE!5ACKO__\$#2???OW?_OS??K???GO\$#3!4?_oWGGKK!4?_\$#5!7?_???OGG\$#7!8?!4_oo\$#6!9?OO???o\$#4!12?C-#6??Wc!4?A!6?W\$#1?}f?A???c?Ac???A?B}\$#2???A?@?W?A??W?@c?c\$#3???@@???W{{W???@~W\$#7???WW{~B!4@B}{\$#5!4?cA?c???Ac@A-#5!5?@?C???HOQ@\$#1?@FC@o__??@??_oHCF@\$#2???HWACGg@?o??AOG\$#3???AEKWOOoo!4?EB\$#7!6?BB!4ENL\$#6!8?@GG???K\$#4!12?_-#1!7?!6@\e\\\e[2C \[\e]0;\u@\h: \w\a\]${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '
```

![image](https://github.com/user-attachments/assets/4e9ab318-43f8-4585-ac82-8350cb670dac)

This can be set in your `~/.bashrc` file to do it automatically when launching bash. For exemple by using the command line `echo "${PS1@A}" >> ~/.bashrc` when you are satisfied with your current prompt.

Note however that some features relying on Bash computing the width of the prompt will get slightly confused, such as the tab completion, and up/down history, when it believes the command has reached the end of the line.

---

Note all of these require a version of Windows Terminal that supports sixels. At the time of this writing, that means Windows Terminal Preview 1.22.2362.0, as the non-preview version is still on 1.20 which does not support sixels.

**Comment by trinitronx** [comment]
Maybe deserves a [PR here][1] after release to update https://www.arewesixelyet.com/

[1]: https://github.com/smasher164/arewesixelyet

**Reply by j4james** [reply]
I don't think that website is maintained anymore. There is already an outstanding PR for Windows Terminal (multiple PRs actually), but nothing has been merged in more than a year.
